doc/developer/design/20230421_stabilize_monotonic_select.md

```diff
@@ -0,0 +1,137 @@
+
+# Reference explanation
+[reference-explanation]: #reference-explanation
+
+In the following, we consider the main outstanding issues to stabilize monotonic one-shot `SELECT`s and propose strategies to address each in turn.
+
+## Revisiting the Output of `EXPLAIN` (MaterializeInc/materialize#18089)
+
+A monotonic one-shot `SELECT` has a physical plan that differs from the one for the same `SELECT` in an indexed or materialized view. We therefore propose that syntax be provided to enable `EXPLAIN`, and in particular `EXPLAIN PHYSICAL PLAN`, to either produce one-shot `SELECT` plans or plans for other optimization paths.
+
+Issue MaterializeInc/materialize#18089 captures the discussion and design of the syntax for `EXPLAIN` to reflect different optimization paths. We defer the proposal of a concrete syntax to the design originating from that issue.
+
+Given an adequate syntax, the optimization path for one-shot `SELECT`s in `EXPLAIN` should then include a setting of both `as_of` and `until` prior to invoking `finalize_dataflow`. It is critical that `until` be set to `as_of + 1`, assuming both attributes refer to a single-timestamp, one-dimensional frontier. This proper setting of `as_of` and `until` was already implemented in MaterializeInc/materialize#19223.
+
+## Running Tests in CI with Both Monotonic and Non-Monotonic Plan Variants (MaterializeInc/materialize#18734)
+
+Reusing `sqllogictest` corpora allows us to get a high coverage of SQL constructs. So it is ideal that we be able to continue reusing corpora designed for other DBMS with low effort, even after the feature flag `enable_monotonic_oneshot_selects` is retired. At the same time, other DBMS may not introduce the same differentiation between one-shot `SELECT` plans and plans for indexed or materialized views as in Materialize. Therefore, their corpora would start to fundamentallly lack coverage of the plans in Materialize that are specialized to view maintenance, once the feature of monotonic one-shot `SELECT`s is enabled.
+
+We propose to introduce a mode in our `sqllogictest` runner wherein a `Query` directive is run in two different ways. Firstly, the query is executed as a normal one-shot `SELECT`, as done now in `run_query`. Secondly, we issue a `CREATE VIEW` for the query SQL followed by a `CREATE DEFAULT INDEX` on the view. The query is then executed again as a `SELECT * FROM v`, where `v` is the view name provided. The outputs of the two executions are compared against the reference output.
+
+Automatically created view names above should be generated so as to avoid clashes with view names defined in tests. A simple strategy is, for example, to include a UUID in the view name string. Additonally, the special mode referred to above can be invoked by providing an additional option to the `sqllogictest` executable, e.g., `--auto_index_selects`, inspired by the existing option `auto_index_tables`.
```

Comment:
Is the proposal to run all sqllogictests once with and once without `--auto_index_selects`, or only a relevant subset? If we run all tests then runtime might be a concern. We already had in the past issues with the full SLT run exceeding the 10-hour timeout. This might become an issue again, especially because we would be adding a lot of new DDL statements which seem to be significantly slower than DML in Materialize.

Reply:
I am personally OK with the sqllogic-level rewrites , as other databases are also doing this (MySQL's MTR test runner comes to mind). However, in our case we do expect different plans for the one-shot and materialized version of the query. Therefore, the rewrite should be something like this:

```
EXPLAIN SELECT ...
```
gets rewritten to:



```
EXPLAIN SELECT ... # This one is kept
CREATE MATERIALIZED VIEW ... AS SELECT
EXPLAIN MATERIALIZED VIEW ... # This one is not implemented yet
```

Would it be possible to move the implementation of `EXPLAIN MATERIALIZED VIEW` ?

Reply:
> Is the proposal to run all sqllogictests once with and once without `--auto_index_selects`, or only a relevant subset? If we run all tests then runtime might be a concern. We already had in the past issues with the full SLT run exceeding the 10-hour timeout. This might become an issue again, especially because we would be adding a lot of new DDL statements which seem to be significantly slower than DML in Materialize.

Perhaps we can evaluate where the runtime cost is too large and devise a subset selection strategy appropriately?

There are (at least) the following scenarios to consider: (1) The CI run for each PR; (2) The nightly CI run; (3) The slow SLT pipeline. For (1), we may be able to get away with running SLTs twice, but it might be better to narrow down to a subset, e.g., by pattern-based selection as outlined in the alternatives.

For (2), one primary issue is that SQLancer / SQLsmith would be primarily run on the monotonic one-shot plan variants. I am not sure whether we should aim to get higher coverage also of fuzzing-based tests prior to stabilizing monotonic one-shot `SELECT`s?

For (3), it is highly likely that we would run into troubles with the total time needed to execute these tests. A first strategy would be to try pattern-based selection of a subset; if the time would still be too large, we could refine the subset selection by sampling. The latter can be follow a random or just a simple round-robin strategy. Here, round-robin has an advantage in that every SLT would eventually get a chance to run over a given number of repetitions of the pipeline, but it requires to save state or implement a deterministic mapping (e.g., tests to day-of-week or something).

Reply:
WRT the performance impact, you can place the Buildkite step that performs the rewritten execution of SLT in the Nightly.